File: src/main/java/com/github/shatteredsuite/scrolls/items/NBTUtils.java
package com.github.shatteredsuite.scrolls.items;

import com.github.shatteredsuite.core.include.nbt.NBTCompound;
import com.github.shatteredsuite.core.include.nbt.NBTItem;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;

public class NBTUtils {

    public static Location locationFromNBTItem(NBTItem compound, String offset) {
        double x = compound.getDouble(offset + "x");
        double y = compound.getDouble(offset + "y");
        double z = compound.getDouble(offset + "z");
        float yaw = compound.getFloat(offset + "yaw");
        float pitch = compound.getFloat(offset + "pitch");
        String worldName = compound.getString(offset + "world");
        World world = Bukkit.getWorld(worldName);
        return new Location(world, x, y, z, yaw, pitch);
    }

    public static Location locationFromNBTCompound(NBTCompound compound, String offset) {
        double x = compound.getDouble(offset + "x");
        double y = compound.getDouble(offset + "y");
        double z = compound.getDouble(offset + "z");
        float yaw = compound.getFloat(offset + "yaw");
        float pitch = compound.getFloat(offset + "pitch");
        String worldName = compound.getString(offset + "world");
        World world = Bukkit.getWorld(worldName);
        return new Location(world, x, y, z, yaw, pitch);
    }

    public static void addLocationNBT(NBTCompound compound, Location location) {
        compound.setDouble("x", location.getX());
        compound.setDouble("y", location.getY());
        compound.setDouble("z", location.getZ());
        compound.setFloat("yaw", location.getYaw());
        compound.setFloat("pitch", location.getPitch());
        if(location.getWorld() != null) {
            compound.setString("world", location.getWorld().getName());
        }
        else {
            compound.setString("world", Bukkit.getWorlds().get(0).getName());
        }
    }
}

File: src/main/kotlin/com/github/shatteredsuite/scrolls/items/NBTUtils.kt
package com.github.shatteredsuite.scrolls.items

import com.github.shatteredsuite.core.include.nbt.NBTCompound
import com.github.shatteredsuite.core.include.nbt.NBTItem
import org.bukkit.Bukkit
import org.bukkit.Location
import java.util.*

object NBTUtils {
    fun locationFromNBTItemUUID(compound: NBTItem, offset: String): Location {
        val x = compound.getDouble(offset + "x")
        val y = compound.getDouble(offset + "y")
        val z = compound.getDouble(offset + "z")
        val yaw = compound.getFloat(offset + "yaw")
        val pitch = compound.getFloat(offset + "pitch")
        val worldName = compound.getString(offset + "world")
        val world = Bukkit.getWorld(UUID.fromString(worldName))
        return Location(world, x, y, z, yaw, pitch)
    }

    fun locationFromNBTItem(compound: NBTItem, offset: String): Location {
        val x = compound.getDouble(offset + "x")
        val y = compound.getDouble(offset + "y")
        val z = compound.getDouble(offset + "z")
        val yaw = compound.getFloat(offset + "yaw")
        val pitch = compound.getFloat(offset + "pitch")
        val worldName = compound.getString(offset + "world")
        val world = Bukkit.getWorld(worldName)
        return Location(world, x, y, z, yaw, pitch)
    }

    fun locationFromNBTCompound(compound: NBTCompound, offset: String): Location {
        val x = compound.getDouble(offset + "x")
        val y = compound.getDouble(offset + "y")
        val z = compound.getDouble(offset + "z")
        val yaw = compound.getFloat(offset + "yaw")
        val pitch = compound.getFloat(offset + "pitch")
        val worldName = compound.getString(offset + "world")
        val world = Bukkit.getWorld(worldName)
        return Location(world, x, y, z, yaw, pitch)
    }

    fun addLocationNBT(compound: NBTCompound, location: Location) {
        compound.setDouble("x", location.x)
        compound.setDouble("y", location.y)
        compound.setDouble("z", location.z)
        compound.setFloat("yaw", location.yaw)
        compound.setFloat("pitch", location.pitch)
        if (location.world != null) {
            compound.setString("world", location.world!!.name)
        } else {
            compound.setString("world", Bukkit.getWorlds()[0].name)
        }
    }
}